### PR TITLE
[OpenCL] Add Command Buffer extension to OpenCL adapter.

### PR DESCRIPTION
**Pull request by martygrant**
Adds implementations for the [experimental Command Buffer](https://oneapi-src.github.io/unified-runtime/core/api.html#command-buffer-experimental) entry points. Also fixes minor build error with missing namespace in enqueue.cpp.

Added companion intel-llvm PR which also updates many of the E2E tests to run against OpenCL - https://github.com/intel/llvm/pull/11718

### CODE REVIEW

File: source/adapters/opencl/command_buffer.cpp
Comment (EwanC) [inline source/adapters/opencl/command_buffer.cpp:64]: `Result` for consistent variable naming

File: source/adapters/opencl/command_buffer.cpp
Comment (EwanC) [inline source/adapters/opencl/command_buffer.cpp:17]: `Queue` to be PascalCase

File: source/adapters/opencl/command_buffer.cpp
Comment (EwanC) [inline source/adapters/opencl/command_buffer.cpp:23]: `clContext` should start with a capital for PascalCase variable naming, see this called `CLContext ` in other files. Same comment for `clCommandBuffer` and `urCommandBuffer` later on

File: source/adapters/opencl/command_buffer.cpp
Comment (EwanC) [inline source/adapters/opencl/command_buffer.cpp:33]: Should we be checking the return value from  `Res` here?

File: source/adapters/opencl/command_buffer.hpp
Comment (EwanC) [inline source/adapters/opencl/command_buffer.hpp:14]: Can remove this comment

File: source/adapters/opencl/common.hpp
Comment (EwanC) [inline source/adapters/opencl/common.hpp:239]: Rather than defining these ourselves, can we use the symbols from the OpenCL-Headers instead? https://github.com/KhronosGroup/OpenCL-Headers/blob/main/CL/cl_ext.h#L102-L110

Not sure what version of the headers you're testing, with but these should have been in for a while

File: source/adapters/opencl/command_buffer.cpp
Comment (EwanC) [inline source/adapters/opencl/command_buffer.cpp:7]: Why are we touching these license lines? If this is relevant think it should be a standalone PR that updates all the affected files

Comment (kbenzie) [reply source/adapters/opencl/command_buffer.cpp:7]: Yeah, this license change should be reverted. The removed license comments are the correct ones for UR.

File: source/adapters/opencl/command_buffer.cpp
Comment (martygrant) [reply source/adapters/opencl/command_buffer.cpp]: Fixed, thanks.

File: source/adapters/opencl/command_buffer.cpp
Comment (martygrant) [reply source/adapters/opencl/command_buffer.cpp]: Fixed, thanks.

File: source/adapters/opencl/command_buffer.cpp
Comment (martygrant) [reply source/adapters/opencl/command_buffer.cpp]: Fixed, thanks. Added a CL_RETURN_ON_FAILURE_AND_SET_NULL for `Res`.

File: source/adapters/opencl/command_buffer.hpp
Comment (martygrant) [reply source/adapters/opencl/command_buffer.hpp]: Fixed, thanks.

File: source/adapters/opencl/common.hpp
Comment (martygrant) [reply source/adapters/opencl/common.hpp:239]: Discussed this offline. Leaving this as is to be safe for the Github CI.

File: source/adapters/opencl/command_buffer.cpp
Comment (martygrant) [reply source/adapters/opencl/command_buffer.cpp]: Fixed thanks think this was an accidental copy paste error.

**Review by veselypeta — APPROVED** [review]

File: source/adapters/opencl/command_buffer.cpp
Comment (EwanC) [inline source/adapters/opencl/command_buffer.cpp:324]: Why do we need this check? [urCommandBufferEnqueueExp](https://oneapi-src.github.io/unified-runtime/core/api.html#urcommandbufferenqueueexp) specifies that if `hQueue` is NULL then `UR_RESULT_ERROR_INVALID_NULL_HANDLE` should be returned

Comment (martygrant) [reply source/adapters/opencl/command_buffer.cpp:324]: I added this from reading the below from the [OpenCL spec](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#_add_new_section_section_5_x_2_enqueuing_a_command_buffer).
```
queues A pointer to an ordered list of command-queues [compatible](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_Ext.html#compatible) with the command-queues used on recording. queues can be NULL in which case the default command-queues used on command-buffer creation are used and num_queues must be 0.
```

File: source/adapters/opencl/command_buffer.cpp
Comment (martygrant) [reply source/adapters/opencl/command_buffer.cpp:67]: Yeah that should be called too I'll make sure to add that in, thanks.

File: source/adapters/opencl/command_buffer.cpp
Comment (EwanC) [reply source/adapters/opencl/command_buffer.cpp:324]: Sure, but that OpenCL behaviour isn't a valid implementation of this UR entry-point. For example, if we had a CTS test that tested calling `urCommandBufferEnqueueExp` with a null `hQueue` parameter, then the expected behaviour is to return an error `UR_RESULT_ERROR_INVALID_NULL_HANDLE `, not enqueue the command-buffer.

File: source/adapters/opencl/command_buffer.cpp
Comment (fabiomestre) [inline source/adapters/opencl/command_buffer.cpp:21]: Minor: There is a macro in the adapter called UR_RETURN_ON_FAILURE that could be used to simplify this error checks

File: source/adapters/opencl/command_buffer.cpp
Comment (martygrant) [reply source/adapters/opencl/command_buffer.cpp]: Apologies that's my misunderstanding. I wasn't aware of how the validation layer works before but have now found out those checks specified in the spec are done [in that layer](https://github.com/oneapi-src/unified-runtime/blob/main/source/loader/layers/validation/ur_valddi.cpp#L7434). I've removed the `uint32_t numQueues = (hQueue) ? 1 : 0;`. Cheers.

**Review by EwanC — CHANGES_REQUESTED** [review]
This code LGTM, but I think this PR should have a companion change in DPC++. Once this UR commit get's pulled into DPC++ the SYCL-Graph extension will start reporting support for OpenCL with undocumented behaviour. Also without DPC++ testing, i'm not sure how to use this code path, given that I don't think there's UR CTS testing for experimental features.

What i'd imagine the companion DPC++ PR to cover:
* Add a OpenCL section to the design doc for SYCL-Graphs https://github.com/intel/llvm/blob/sycl/sycl/doc/design/CommandGraph.md#backend-implementation documenting the implementation on cl_khr_command_buffer, and explicitly documenting the types of nodes that aren't currently supported.
* Add opencl to the REQUIRES line of `test-e2e/Graph` tests which now pass with this change.
* Add a `device.get_info<exp_ext::info::device::graph_support>();` check to all tests, skipping returning early for those which don't report support. As even for an OpenCL device, support is dependent on the cl_khr_command_buffer extension being available. This will probably be the case for all DPC++ CI testing, which I don't think uses a device with cl_khr_command_buffer support.

**Comment by martygrant** [comment]
Hey @EwanC, thanks for the really thorough review. That all sounds fine for a new PR into DPC++, I'll open up an issue with the details you've provided. Should one PR be merged before the other?

We're temporarily holding off merging any more stuff into the adapters branch here until a L0 issue is reverted but I'll have both of these changes ready to go once that's sorted.

**Comment by EwanC** [comment]
> Hey @EwanC, thanks for the really thorough review. That all sounds fine for a new PR into DPC++, I'll open up an issue with the details you've provided. Should one PR be merged before the other?
> 
> We're temporarily holding off merging any more stuff into the adapters branch here until a L0 issue is reverted but I'll have both of these changes ready to go once that's sorted.

I think following the existing adapter change process is fine https://oneapi-src.github.io/unified-runtime/core/CONTRIB.html#adapter-change-process This will need to be merged first, but only after the DPC++ PR has been approved and CI passing.

**Comment by martygrant** [comment]
Have opened up an issue for the work required in DPC++ - https://github.com/intel/llvm/issues/11599

**Comment by martygrant** [comment]
Corrected an error in `urCommandBufferAppendKernelLaunchExp` where the `cl_command_queue command_queue` parameter for `clCommandNDRangeKernelKHR` must be NULL. Given the extension function pointers more meaningful names instead of just `F`.

File: source/adapters/opencl/command_buffer.cpp
Comment (EwanC) [inline source/adapters/opencl/command_buffer.cpp:173]: From the spec wording from `clCommandCopyBufferKHR` for where we are passing the internal queue:

> command_queue Specifies the command-queue the command will be recorded to. Parameter is unused by this extension as only a single command-queue is supported and must be NULL.

Similar wording for other entry-points. We should replace this internal queue parameter passed on command-recording entry-points with NULL

File: source/adapters/opencl/command_buffer.cpp
Comment (EwanC) [inline source/adapters/opencl/command_buffer.cpp:137]: Can we remove these die calls? they result in an abort which isn't ideal now that it's highly likely a user could use these entry-points by accident. Returning the error code will be enough for the SYCL runtime to throw an sensible exception that the end user can deal with how they please.

File: source/adapters/opencl/command_buffer.cpp
Comment (martygrant) [reply source/adapters/opencl/command_buffer.cpp]: Fixed: `clCommandCopyBufferKHR`, `clCommandCopyBufferRectKHR` and `clCommandFillBufferKHR`

**Review by EwanC — APPROVED** [review]
LGTM 😃

**Pull request merged**
